File: src/engine/knowledgeGraph.js
// Layer 1: hand-authored DAG of math skills. Transcribed from spec §4.
// Note: spec summary says "12 nodes" but the §4 table lists 13; we implement all 13
// and derive any model dimensions from SKILL_IDS.length (so the DKT input dim is
// 2 * SKILL_IDS.length = 26, not the spec's hard-coded 24).

export const SKILLS = {
  'counting':        { description: 'Number recognition, ordering', grade: '2' },
  'addition':        { description: 'Single & multi-digit addition', grade: '2-3' },
  'subtraction':     { description: 'Single & multi-digit subtraction', grade: '2-3' },
  'multiplication':  { description: 'Times tables, multi-digit', grade: '3-4' },
  'division':        { description: 'Basic division, remainders', grade: '4-5' },
  'patterns':        { description: 'Sequences, AP/GP basics', grade: '3-5' },
  'fractions-basic': { description: 'Identifying & comparing fractions', grade: '4-5' },
  'equiv-fractions': { description: 'Equivalence, addition of fractions', grade: '5-6' },
  'decimals':        { description: 'Decimal operations', grade: '5-6' },
  'integers':        { description: 'Negative numbers', grade: '5-6' },
  'geometry-shapes': { description: 'Shapes, angles, area, perimeter', grade: '4-6' },
  'coord-geometry':  { description: 'Coordinate plane, distance', grade: '6+' },
  'algebra-basics':  { description: 'Variables, simple equations', grade: '6+' },
};

export const SKILL_IDS = Object.keys(SKILLS);

// prereq -> the skill cannot be attempted until these are mastered (spec §4 DAG).
// The "(equations)" label on the algebra->geometry edge in the §4 diagram is not a
// skill; it is collapsed into the algebra-basics -> geometry-shapes edge.
const PREREQS = {
  'counting':        [],
  'addition':        ['counting'],
  'subtraction':     ['addition'],
  'multiplication':  ['subtraction'],
  'division':        ['multiplication'],
  'patterns':        ['addition', 'subtraction'],
  'integers':        ['multiplication'],
  'fractions-basic': ['division'],
  'equiv-fractions': ['fractions-basic'],
  'decimals':        ['fractions-basic'],
  'coord-geometry':  ['decimals'],
  'algebra-basics':  ['patterns'],
  'geometry-shapes': ['algebra-basics'],
};

// game page (component name) -> skills exercised (spec §4 table).
export const GAME_SKILLS = {
  ArithmeticGame:       ['addition', 'subtraction'],
  MultiplicationMeteor: ['multiplication'],
  MultiplicationFarm:   ['multiplication'],
  FractionFrenzy:       ['fractions-basic'],
  FractionNinja:        ['fractions-basic', 'equiv-fractions'],
  EquationBalancer:     ['algebra-basics'],
  AlgebraDungeon:       ['algebra-basics'],
  GeometryGame:         ['geometry-shapes'],
  CoordinateTreasure:   ['coord-geometry'],
  DecimalMall:          ['decimals'],
  IntegerMountain:      ['integers'],
  PatternPuzzle:        ['patterns'],
  NumberCatcher:        ['counting', 'patterns'],
  BalloonPopSequence:   ['counting', 'patterns'],
  FruitRush:            ['addition', 'multiplication'],
  MathRacing:           ['addition', 'multiplication'],
};

export function getPrereqs(skillId) {
  return PREREQS[skillId] ?? [];
}

export function arePrereqsMet(skillId, mastery, cutoff = 0.75) {
  return getPrereqs(skillId).every((p) => (mastery[p] ?? 0) >= cutoff);
}

// children[skill] = skills that list `skill` as a prerequisite.
const CHILDREN = SKILL_IDS.reduce((acc, id) => {
  acc[id] = SKILL_IDS.filter((other) => getPrereqs(other).includes(id));
  return acc;
}, {});

export function getDescendants(skillId) {
  const seen = new Set();
  const stack = [...(CHILDREN[skillId] ?? [])];
  while (stack.length) {
    const cur = stack.pop();
    if (seen.has(cur)) continue;
    seen.add(cur);
    stack.push(...(CHILDREN[cur] ?? []));
  }
  return [...seen];
}

export function getLeverage(skillId) {
  return getDescendants(skillId).length;
}

export function getGamesForSkill(skillId) {
  return Object.keys(GAME_SKILLS).filter((g) => GAME_SKILLS[g].includes(skillId));
}

// Kahn's algorithm — throws if the graph has a cycle.
export function topologicalOrder() {
  const indeg = {};
  for (const id of SKILL_IDS) indeg[id] = getPrereqs(id).length;
  const queue = SKILL_IDS.filter((id) => indeg[id] === 0);
  const order = [];
  while (queue.length) {
    const node = queue.shift();
    order.push(node);
    for (const child of CHILDREN[node]) {
      indeg[child] -= 1;
      if (indeg[child] === 0) queue.push(child);
    }
  }
  if (order.length !== SKILL_IDS.length) {
    throw new Error('knowledgeGraph: prerequisite cycle detected');
  }
  return order;
}

File: scripts/eval/knowledge_graph.py
"""Python re-encoding of src/engine/knowledgeGraph.js (the JS source is canonical).

13 skills, prerequisite DAG, plus the decision-layer constants the eval harness needs.
Kept manually in sync; tests/test_knowledge_graph.py guards against drift.
"""
from __future__ import annotations

# Ordered exactly as Object.keys(SKILLS) in knowledgeGraph.js — this order defines
# every skill_idx / one-hot index used in the trajectory dataset.
SKILL_IDS: list[str] = [
    "counting",
    "addition",
    "subtraction",
    "multiplication",
    "division",
    "patterns",
    "fractions-basic",
    "equiv-fractions",
    "decimals",
    "integers",
    "geometry-shapes",
    "coord-geometry",
    "algebra-basics",
]

NUM_SKILLS: int = len(SKILL_IDS)  # 13
DKT_INPUT_DIM: int = 2 * NUM_SKILLS  # 26 — see knowledgeGraph.js header note

# prereq -> skills that must be mastered first (transcribed from PREREQS in the JS).
PREREQS: dict[str, list[str]] = {
    "counting": [],
    "addition": ["counting"],
    "subtraction": ["addition"],
    "multiplication": ["subtraction"],
    "division": ["multiplication"],
    "patterns": ["addition", "subtraction"],
    "integers": ["multiplication"],
    "fractions-basic": ["division"],
    "equiv-fractions": ["fractions-basic"],
    "decimals": ["fractions-basic"],
    "coord-geometry": ["decimals"],
    "algebra-basics": ["patterns"],
    "geometry-shapes": ["algebra-basics"],
}

# Decision-layer constants mirrored from decisionLayer.js.
MASTERY_CUTOFF: float = 0.75   # "mastered" for unlock/prereq/breadth
PREREQ_LEARN_GATE: float = 0.5  # spec §5.3: cannot LEARN a skill until prereqs > 0.5

_SKILL_INDEX = {sid: i for i, sid in enumerate(SKILL_IDS)}


def skill_index(skill_id: str) -> int:
    return _SKILL_INDEX[skill_id]


def get_prereqs(skill_id: str) -> list[str]:
    return PREREQS.get(skill_id, [])


def are_prereqs_met(skill_id: str, mastery: dict[str, float], cutoff: float = MASTERY_CUTOFF) -> bool:
    return all(mastery.get(p, 0.0) >= cutoff for p in get_prereqs(skill_id))


def prereqs_learnable(skill_id: str, mastery: dict[str, float], gate: float = PREREQ_LEARN_GATE) -> bool:
    """spec §5.3 learn gate: a skill cannot be LEARNED until every prereq > `gate`."""
    return all(mastery.get(p, 0.0) > gate for p in get_prereqs(skill_id))


# children[skill] = skills that list `skill` as a prerequisite.
_CHILDREN: dict[str, list[str]] = {
    sid: [other for other in SKILL_IDS if sid in get_prereqs(other)] for sid in SKILL_IDS
}


def get_children(skill_id: str) -> list[str]:
    return _CHILDREN.get(skill_id, [])


def get_descendants(skill_id: str) -> list[str]:
    seen: set[str] = set()
    stack = list(_CHILDREN.get(skill_id, []))
    while stack:
        cur = stack.pop()
        if cur in seen:
            continue
        seen.add(cur)
        stack.extend(_CHILDREN.get(cur, []))
    return sorted(seen)


def get_leverage(skill_id: str) -> int:
    return len(get_descendants(skill_id))


def topological_order() -> list[str]:
    """Kahn's algorithm — raises on a cycle."""
    indeg = {sid: len(get_prereqs(sid)) for sid in SKILL_IDS}
    queue = [sid for sid in SKILL_IDS if indeg[sid] == 0]
    order: list[str] = []
    while queue:
        node = queue.pop(0)
        order.append(node)
        for child in _CHILDREN[node]:
            indeg[child] -= 1
            if indeg[child] == 0:
                queue.append(child)
    if len(order) != len(SKILL_IDS):
        raise ValueError("knowledge_graph: prerequisite cycle detected")
    return order
